devito/types/tensor.py

```diff
@@ -0,0 +1,598 @@
+from collections import OrderedDict
+
+import sympy
+import numpy as np
+from sympy.core.sympify import converter as sympify_converter
+from sympy.core.decorators import call_highest_priority
+
+from devito.finite_differences import Differentiable
+from devito.logger import error
+from devito.types.utils import NODE
+from devito.types.basic import AbstractCachedTensor
+from devito.types.dense import Function, TimeFunction
+
+__all__ = ['TensorFunction', 'TensorTimeFunction', 'VectorFunction', 'VectorTimeFunction']
+
+
+class TensorFunction(AbstractCachedTensor, Differentiable):
+    """
+    Tensor valued Function represented as an ImmutableMatrix.
```

Comment:
don't think you need to mention ImmutableMatrix here. Users won't digest it. And this is a docstring for users